Automattic/wp-calypso
# Redux dev: Split actionLog and consoleDispatch enhancers

## The issue

The console dispatcher redux store enhancers serves several purposes.

* It attaches the top-level redux store methods to the window object so they are accessible for development, including `dispatch` as its name implies
* It adds some logging for dispatched actions and exposes some operations on the logged actions by attaching them to the window object.

These are two separate concerns, and including them in a single store enhancer is not feasible:
* Console dispatching must happen near the top of the enhancer stack so that they are visible to other store enhancers, especially our app middleware. This is OK in the current implementation.
* Action logging must happen near the bottom of the enhancer stack so that it has visibility on the actions dispatch by other enhancers, especially our app middleware. This is broken in the current implementation, middleware dispatched actions are invisible to the console action logger.

## The proposed fix

This PR solves the issue with missing actions from `window.actionLog` by splitting the existing enhancer into two and placing the "logging" enhancer near the bottom of the stack:

* Console Dispatcher - This enhancer keeps its name and position at the top of the enhancer stack. It's implementation is vastly simplified as its only responsibility at this point is to expose the store methods on the `window` object if it exists.
* Action Logger - All code related to the action logger has moved from the Console Dispatcher into this enhancer. It has been added near the bottom of the enhancer stack, just before the enhancer for the browser extension redux devtools.

## Context

This issue was discovered while testing https://github.com/Automattic/wp-calypso/pull/20638#issuecomment-350991262

## Testing
1. `dispatch( { type: actionTypes.USER_PROFILE_LINKS_REQUEST } )`
1. Ensure network request fires as expected
1. `actionLog.filter( actionTypes.USER_PROFILE_LINKS_REQUEST )`
1. Ensure you see the request action as expected
1. `actionLog.filter( actionTypes.USER_PROFILE_LINKS_RECEIVE )`
1. Ensure you see the receive action as expected
1. Ensure other [redux store methods](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#store-methods) continue to behave as expected (no changes):
   * [`getState()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#getstate) / `state`: provides access to the current redux state
   * [`dispatch()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#dispatchaction): see above
   * [`subscribe()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#subscribelistener)
   * [`replaceReducer()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#replacereducernextreducer)

You can try the steps mentioned in https://github.com/Automattic/wp-calypso/pull/20638#issuecomment-350991262 to ensure the fix works.
